### PR TITLE
[WIP] qrexec internals diagrams

### PR DESCRIPTION
**Pull request by pierwill**
:wave: 

Before revising the qrexec internals docs themselves, I want to create diagrams for the two basic use cases: [invoking an RPC command from dom0](https://www.qubes-os.org/doc/qrexec-internals/#dom0-request-execution-of-some_command-in-domx-and-pass-stdinstdout), and [invoking from domY](https://www.qubes-os.org/doc/qrexec-internals/#domy-invoke-execution-of-qubes-service-qubessomerpc-in-domx-and-pass-stdinstdout).

I've used the current documentation to sketch out the general shape of the protocol interactions. There is definitely some guesswork here---but keep in mind that these diagrams aren't meant to include all details. For instance, they only show successfully completed qrexec calls! Also, there are comments in the mermaid source that could be included in the paragraphs around the diagrams.

These links will take you to a code playground at the mermaidjs site where you can see the drafts so far:

- [dom0 RPC](https://mermaidjs.github.io/mermaid-live-editor/#/edit/eyJjb2RlIjoic2VxdWVuY2VEaWFncmFtXG5wYXJ0aWNpcGFudCAwZCBhcyBkb20wIHFyZXhlYy1kYWVtb25cbnBhcnRpY2lwYW50IDBjIGFzIGRvbTAgcXJleGVjLWNsaWVudFxucGFydGljaXBhbnQgeGQgYXMgZG9tWCBxcmV4ZWMtZGFlbW9uXG5wYXJ0aWNpcGFudCB4YSBhcyBkb21YIHFyZXhlYy1hZ2VudFxuIyBxcmV4ZWMtY2xpZW50IC1kIGRvbVggdXNlcjpzb21lX2NvbW1hbmRcbjBjIC0tPj4geGQ6IGNvbm5lY3Rpb24gc3RhcnRlZFxuMGMgLT4-IHhkOiBNU0dfSEVMTE8gaGVhZGVyICsgcGVlcl9pbmZvXG4wYyAtPj4gMGQ6IE1TR19IRUxMTyBoZWFkZXIgKyBwZWVyX2luZm9cbjBjIC0-PiB4ZDogTVNHX0VYRUNfQ01ETElORSArIGV4ZWNfcGFyYW1zXG54ZCAtPj4gMGM6IE1TR19FWEVDX0NNRExJTkUgKyBleGVjX3BhcmFtcyAoZW1wdHkgY21kKVxuMGQgLT4-IHhhOiBNU0dfRVhFQ19DTURMSU5FICsgZXhlY19wYXJhbXNcbjBjIC0teCB4ZDogZGlzY29ubmVjdHNcbiMgTm90ZSBvdmVyIDBjOiBzdGFydHMgYSB2Y2hhbiBzZXJ2ZXJcbjBjIC0-PiAwYzogdmNoYW4gc2VydmVyIHdhaXRzIGZvciBjb25uZWN0aW9uXG4wZCAtPj4geGE6IE1TR19FWEVDX0NNRExJTkUgKyBleGVjX3BhcmFtcyAob3ZlciB2Y2hhbilcbnhhIC0tPj4gMGM6IGNvbm5lY3RzIG92ZXIgdmNoYW4gdXNpbmcgZXhlY19wYXJhbXNcbk5vdGUgb3ZlciB4YTogZXhlY3V0ZXMgc29tZV9jb21tYW5kXG54YSAtPj4gMGM6IHN0ZGluL3N0b3V0IChvdmVyIHZjaGFuKVxuTm90ZSBvdmVyIHhhOiBzb21lX2NvbW1hbmQgdGVybWluYXRlc1xueGEgLT4-IDBjOiBNU0dfREFUQV9FWElUX0NPREVcbnhhIC0teCAwYzogZGlzY29ubmVjdHMiLCJtZXJtYWlkIjp7fX0)
- [domY RPC](https://mermaidjs.github.io/mermaid-live-editor/#/edit/eyJjb2RlIjoic2VxdWVuY2VEaWFncmFtXG5wYXJ0aWNpcGFudCB5ZCBhcyBxcmV4ZWMtZGFlbW9uIChkb21ZKVxucGFydGljaXBhbnQgeGQgYXMgcXJleGVjLWRhZW1vbiAoZG9tWClcbnBhcnRpY2lwYW50IDBjIGFzIHFyZXhlYy1jbGllbnRcbnBhcnRpY2lwYW50IHlhIGFzIGRvbVkgcXJleGVjLWFnZW50XG5wYXJ0aWNpcGFudCB4YSBhcyBkb21YIHFyZXhlYy1hZ2VudFxuIyBxcmV4ZWMtY2xpZW50LXZtIGludm9rZWQsIHFyZXhlYy1hZ2VudCBnZW5lcmF0ZXMgcmVxX2lkXG55YSAtPj4geWQ6IG1zZyB0cmlnZ2VyIHNlcnZpY2UgcGFyYW1zICsgcmVxX2lkXG5Ob3RlIG92ZXIgeWQ6IGNoZWNrcyBwb2xpY3lcbnlkIC0-PiAwYzogZXhpdCBjb2RlXG4wYyAtLT4-IHhkOiAoaWYgcG9saWN5IGFsbG93cykgY29ubmVjdHNcbnhkIC0-PiAwYzogTVNHX0hFTExPICsgcGVlcl9pbmZvXG4wYyAtPj4geGQ6IE1TR19IRUxMTyArIHBlZXJfaW5mb1xuMGMgLT4-IHhkOiBNU0dfRVhFQ19DTURMSU5FICsgZXhlY19wYXJhbXNcbnhkIC0-PiAwYzogTVNHX0VYRUNfQ01ETElORSArIGV4ZWNfcGFyYW1zLCBlbXB0eSBjbWRcbnhkIC0-PiB4YTogTVNHX0VYRUNfQ01ETElORSArIGV4ZWNfcGFyYW1zXG4wYyAtLXggeGQ6IGRpc2Nvbm5lY3RzXG4wYyAtLT4geWQ6IFtjb25uZWN0aW9uIGFuZCBpbml0aWFsIE1TR19IRUxMTyBleGNoYW5nZV1cbjBjIC0-PiB5ZDogTVNHX1NFUlZJQ0VfQ09OTkVDVCArIGV4ZWNfcGFyYW1zIChjbWQgPSByZXFfaWQpXG55ZCAtPj4geWE6IE1TR19TRVJWSUNFX0NPTk5FQ1QgKyBleGVjX3BhcmFtcyAoY21kID0gcmVxX2lkKVxueWEgLT4-IHlhOiB2Y2hhbiBzZXJ2ZXIgd2FpdHNcbnhhIC0tPiB5YTogY29ubmVjdGlvbiBlc3RhYmxpc2hlZFxuIyBxcmV4ZWMtYWdlbnQgbm93IHBsYXlzIHJvbGUgb2YgcXJleGVjLWNsaWVudCBpbiBkb20wLWJhc2VkIGV4YW1wbGUgKHNlZSBhYm92ZSlcbiIsIm1lcm1haWQiOnsidGhlbWUiOiJkZWZhdWx0In19)

Feedback would be greatly appreciated! :pray:

### CODE REVIEW

**Comment by marmarek** [comment]
Dom0 has no separate `qrexec-daemon` instance for itself. Only one for each domU. Updated dom0 RPC diagram:
```
sequenceDiagram
participant 0c as dom0 qrexec-client
participant xd as domX qrexec-daemon
participant xa as domX qrexec-agent
# qrexec-client -d domX user:some_command
0c -->> xd: connection started
0c ->> xd: MSG_HELLO header + peer_info
xd ->> 0c: MSG_HELLO header + peer_info
# allocate and fill vchan port
0c ->> xd: MSG_EXEC_CMDLINE + exec_params
xd ->> 0c: MSG_EXEC_CMDLINE + exec_params (empty cmd)
xd ->> xa: MSG_EXEC_CMDLINE + exec_params
0c --x xd: disconnects
# Note over 0c: starts a vchan server
0c ->> 0c: vchan server waits for connection
xa -->> 0c: connects over vchan using exec_params
Note over xa: executes some_command
xa ->> 0c: stdin/stout (over vchan)
Note over xa: some_command terminates
xa ->> 0c: MSG_DATA_EXIT_CODE
xa --x 0c: disconnects
```

Have I mentioned I love this tool?

**Comment by pierwill** [comment]
How do these look? I'd like to get these in a useable form before tackling the internals rewrite (Phase 3 of the qrexec project).

**Comment by marmarek** [comment]
"exit code" on the arrow to qrexec-client looks weird. It should be exit code of qrexec-policy going to qrexec-daemon(domY), in parallel with starting qrexec-client (if policy allows). I think it would be accurate enough if you remove that arrow and make "qrexec-daemon(domY) -->> qrexec-client: if policy allows" instead.

Otherwise looks good.

File: wiki/qrexec3/qrexec-internals-domY.mmd [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+participant yd as qrexec-daemon (domY)
+participant xd as qrexec-daemon (domX)
+participant 0c as qrexec-client
+participant ya as domY qrexec-agent
+participant xa as domX qrexec-agent
+# qrexec-client-vm invoked, qrexec-agent generates req_id
+ya ->> yd: msg trigger service params + req_id
+Note over yd: checks policy
+yd -->> 0c: (if policy allows) start qrexec
+0c -->> xd: (if policy allows) connection starts
```
Comment (marmarek) [inline wiki/qrexec3/qrexec-internals-domY.mmd]: Now redundant:
```suggestion
0c -->> xd: connection starts
```

**Pull request merged**